client/__tests__/e2e/e2e.test.ts

```diff
@@ -1017,6 +1035,80 @@ for (const testDataset of testDatasets) {
         });
       });
     }
+
+    describe(`Image Download`, () => {
+      async function navigateToAndSnapshotImage(
+        page: Page,
+        info: TestInfo,
+        path: string
+      ) {
+        const imageFile = await fs.readFile(path, { encoding: "base64" });
+
+        // attach the image at path to the dom so we can snapshot it
+        // ensure that the image is rendered on top of the graph
+        await page.evaluate((imgData) => {
+          const img = document.createElement("img");
+          img.id = "downloaded-image";
+          img.src = `data:image/png;base64,${imgData}`;
+          img.style.height = "100vh";
+          img.style.zIndex = "1000";
+          img.style.background = "white";
+          img.style.position = "absolute";
+          img.style.top = "0";
+          document.body.appendChild(img);
+        }, imageFile);
+        await takeSnapshot(page, info);
+
+        // remove the image from the dom
+        await page.evaluate(() => {
+          const img = document.getElementById("downloaded-image");
+          img?.parentNode?.removeChild(img);
+        });
+      }
+      async function downloadAndSnapshotImage(page: Page, info: TestInfo) {
+        const graphPromise = page.waitForEvent("download");
+        await page.getByTestId("download-graph-button").click();
+        const graphDownload = await graphPromise;
+
+        const tmp = os.tmpdir();
+        // make the title safe for file system
+        const safeTitle = info.title.replace(/[^a-z0-9]/gi, "_");
+        const graphPath = `${tmp}/${safeTitle}/${graphDownload.suggestedFilename()}`;
+
+        let legendPromise;
+        if (info.title.includes("categorical")) {
+          // this will capture a separate legend download if the colorby is categorical
+          legendPromise = page.waitForEvent("download");
+        }
+        await graphDownload.saveAs(graphPath);
+
+        if (legendPromise) {
+          const legendDownload = await legendPromise;
+          const legendPath = `${tmp}/${safeTitle}/${legendDownload.suggestedFilename()}`;
+          await legendDownload.saveAs(legendPath);
+
+          await navigateToAndSnapshotImage(page, info, legendPath);
+        }
+        await navigateToAndSnapshotImage(page, info, graphPath);
```

Comment:
Do you think we can combine the `if (info.title.includes("categorical")) {` and `if (legendPromise) {` blocks like so?

```suggestion      
        await graphDownload.saveAs(graphPath);

        if (info.title.includes("categorical")) {
          const legendDownload = await page.waitForEvent("download");
          const legendPath = `${tmp}/${safeTitle}/${legendDownload.suggestedFilename()}`;
          await legendDownload.saveAs(legendPath);

          await navigateToAndSnapshotImage(page, info, legendPath);
        }
        await navigateToAndSnapshotImage(page, info, graphPath);
```

Reply:
I'm going to leave it as is for now. Playwright's Download API isn't well documented, especially in multiple download cases. To be clear, this is the functionality we're watching:

1.  Download listener is initiated
2.  Click button that results in image capture and download
3.  First download is initiated and captured by listener
4.  We prepare a new download listener
5.  Playwright saves the download using `.saveAs`, my understanding is  this is similar, if not exactly, to the behavior of choosing where to download the file in the modal
6.  Once the first download is confirmed and saved, the second download (legend) is fired
7.  Playwright captures download step 5 repeats

I'm afraid that if we created and awaited the download listener after step 5, it would timeout as we missed the download event

Reply:
my 2 cents on this - I believe Timmy's suggested edit cleans it up and simplifies it - not sure we need the extra `if (legendPromise)` but also understand your cautiousness around not missing the `download` event. 

Can we try to just trigger both downloads based on the first `const graphPromise = page.waitForEvent("download");` as we are repeating this await for the same event? 

Then move this to after the condition block of the legend/categorical:

```suggestion
        if (info.title.includes("categorical")) {
          // this will capture a separate legend download if the colorby is categorical
          legendPromise = page.waitForEvent("download"); // is this needed?
        }
     
        if (legendPromise) {
          const legendDownload = await legendPromise;
          const legendPath = `${tmp}/${safeTitle}/${legendDownload.suggestedFilename()}`;
          await legendDownload.saveAs(legendPath);

          await navigateToAndSnapshotImage(page, info, legendPath);
        }
        await graphDownload.saveAs(graphPath); // move graph download here
        await navigateToAndSnapshotImage(page, info, graphPath);
```

Reply:
This won't work because the graph download needs to be saved in order for the legend download to be triggered. In your suggestion, the legend await will hang or capture the initial graph download. Step 6 described above is crucial

Reply:
(i'm actually testing this solutions that are being recommended and they don't work)

Reply:
Ok so that's why it was a suggestion because you're the one who's actually running the tests 😁  Then I'd say keep it in the state that works just maybe add a comment explicitly mentioning the need for step 6 in that sequence of creating the second event listener before the first saveAs

Reply:
Ahh got it! Thanks for the explanation, Seve! SGTM 👍